olive/passes/onnx/moe_experts_distributor.py

```diff
@@ -0,0 +1,259 @@
+                _replace_constant_value(producers[concat_iname_1], experts_per_rank)
+
+                break
+
+
+def run(world_size: int, input_filepath: str, output_dirpath: str, debug: bool = False):
+    basename = Path(input_filepath).stem
+    output_dirpath = Path(output_dirpath)
+
+    model = OnnxModel(onnx.load_model(input_filepath))
+    producers = model.output_name_to_node()
+    consumers = model.input_name_to_nodes()
+
+    if debug:
+        OnnxModel.graph_topological_sort(model.model.graph)
+        _dump_graph(model.model, str(output_dirpath / f"{basename}.graph"))
+
+    experts = []
+    num_experts = None
+    for expert in _iterate_experts(model, producers):
+        if debug:
+            pprint.pprint([n.name for n in expert])
+
+        div_node = expert[5]
+        div_node_oname = div_node.output[0]
+        if not num_experts:
+            num_experts = len(consumers[div_node_oname])
+            if (num_experts % world_size) != 0:
+                raise f"""Number of expert paths on node "{div_node.name}" should be a multiple of
+                        input world-size={world_size} but found {len(consumers[div_node_oname])}"""
+        elif len(consumers[div_node_oname]) != num_experts:
+            raise f"""Inconsistent number of expert across layers.
+                    expected={num_experts}, found={len(consumers[div_node_oname])}"""
+
+        experts.append([n.name for n in expert])
+
+    output_filepaths = []
+    for rank in range(0, world_size):
+        rank_model = OnnxModel(onnx.load_model(input_filepath))
+        producers = rank_model.output_name_to_node()
+        consumers = rank_model.input_name_to_nodes()
+        nodes = {node.name: node for node in rank_model.nodes()}
+
+        _create_ranked_model(nodes, producers, consumers, experts, world_size, num_experts, rank)
+        _insert_commop_nodes(rank_model, nodes, experts, world_size)
+        _fix_shapes(rank_model, nodes, producers, consumers, experts, world_size, num_experts, rank)
+
+        rank_model.prune_graph()
+        output_filepath = str(output_dirpath / f"{basename}_{rank:02d}.onnx")
+        rank_model.save_model_to_file(output_filepath)
+        onnx.checker.check_model(rank_model.model)
+        output_filepaths.append(output_filepath)
+
+        if debug:
+            _dump_graph(rank_model.model, str(output_dirpath / f"{basename}_{rank:02d}.graph"))
+
+    return output_filepaths
+
+
+class MoEExpertsDistributor(Pass):
+    """
+    Split the input model (and insert necessary communication operations)
+    to prepare for distributed inferencing.
+    """
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "world_size": PassConfigParam(
+                type_=int,
+                default=2,
+                required=True,
+                description=("Number of GPU nodes to distribute the model for. Must be greater than 1."),
```

Comment:
Could we add validation to raise the unavailable config value? Refer to https://github.com/microsoft/Olive/blob/b11692168ae025d3544dff0348a109dea3a7d24e/olive/passes/snpe/conversion.py#L87

Reply:
Done!